### PR TITLE
Improve tty driver tx on stm32l4

### PR DESCRIPTION
**Pull request by xvuko**

### CODE REVIEW

**Review by agkaminski — CHANGES_REQUESTED** [review]
It's not really fixing any problem, but might be a fix for now.

`keepidle` logic LGTM

File: multi/stm32l4-multi/tty.c [hunk 1]
```diff
@@ -124,9 +127,19 @@ static void tty_irqthread(void *arg)
 			ctx->rxready = 0;
 		}
 
-		if (tty_txready(ctx) && libtty_txready(&ctx->tty_common)) {
-			*(ctx->base + tdr) = libtty_getchar(&ctx->tty_common, NULL);
-			*(ctx->base + cr1) |= (1 << 7);
+		if (libtty_txready(&ctx->tty_common)) {
+			if (tty_txready(ctx)) {
```
Comment (gerard5) [inline multi/stm32l4-multi/tty.c:133]: I noticed, that you have changed the order, is this expected?
I think that inlined `tty_txready(ctx)` should be first, before call to libtty, `libtty_txready(&ctx->tty_common)`, thus:
```suggestion
		if (tty_txready(ctx)) {
			if (libtty_txready(&ctx->tty_common)) {
```
or previous, single `if` line.

**Comment by anglov** [comment]
Not fixing
while true; do echo "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz" > /dev/ttyACM0; sleep 0.2;  done
after short time cause not recoverable:
> libtty-disc: RX OVERRUN!

I know that it don't gonna working properly, but still user input shoudn't cause inrecoverable errors.

Cannot reproduce issue with cutted log now, but worried that it's also hidden now.

File: multi/stm32l4-multi/tty.c (hunk 1 above)
Comment (agkaminski) [reply multi/stm32l4-multi/tty.c:133]: It's ok as is, we need to go to `else` if there is no data in libtty to disable `keepidle`

**Comment by anglov** [comment]
alsp note if you would like to really fix this issues use ISC-83 for it.

File: multi/stm32l4-multi/tty.c (hunk 1 above)
Comment (gerard5) [reply multi/stm32l4-multi/tty.c:133]: I see, that's ok.

File: multi/stm32l4-multi/tty.c [hunk 2]
```diff
@@ -38,6 +38,8 @@
 
 #define TTY_CNT (TTY1 + TTY2 + TTY3 + TTY4 + TTY5)
 
+#define TTY_BUFSIZE 512
```
[the remark this answers is not in the document]
Comment (xvuko) [reply multi/stm32l4-multi/tty.c]: Done

File: multi/stm32l4-multi/tty.c [hunk 3]
```diff
@@ -112,10 +112,13 @@ static int tty_irqHandler(unsigned int n, void *arg)
 static void tty_irqthread(void *arg)
 {
 	tty_ctx_t *ctx = (tty_ctx_t *)arg;
+	uint8_t keptidle = 0;
```
[the remark this answers is not in the document]
Comment (xvuko) [reply multi/stm32l4-multi/tty.c]: Done

[the remark this answers is not in the document]
Comment (xvuko) [reply multi/stm32l4-multi/tty.c]: Done

**Comment by nalajcie** [comment]
Just FYI, the checks failing are due to some problems with Github Actions currently: https://www.githubstatus.com/incidents/cj7gzzj30411

File: multi/stm32l4-multi/tty.c [hunk 4]
```diff
@@ -403,7 +405,7 @@ int tty_init(void)
 		callbacks.set_cflag = tty_setCflag;
 		callbacks.signal_txready = tty_signalTxReady;
 
-		if (libtty_init(&ctx->tty_common, &callbacks, 64) < 0)
+		if (libtty_init(&ctx->tty_common, &callbacks, TTY_BUFSIZE) < 0)
```
[the remark this answers is not in the document]
Comment (xvuko) [reply multi/stm32l4-multi/tty.c]: This was supposed to improve TX. I updated PR title to indicate that.

**Comment by xvuko** [comment]
For now I'd like to submit this as an TX fix.

**Review by anglov — APPROVED** [review]
Ok for me as improvement. Not fixing all issues.

**Comment by agkaminski** [comment]
Ok as a temporary solution, but 512 bytes per buffer is too much IMHO. It will consume 1 KB of memory per tty

**Comment by xvuko** [comment]
> Ok as a temporary solution, but 512 bytes per buffer is too much IMHO. It will consume 1 KB of memory per tty

Using small default value that can be overridden per tty separate for tx and rx would be my solution. For logging having big tx buffer is quite important.

**Pull request merged**